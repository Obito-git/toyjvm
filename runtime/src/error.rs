use crate::ClassId;
use crate::class_loader::ClassLoaderErr;
use crate::rt::LinkageError;
use crate::rt::constant_pool::error::RuntimePoolError;
use common::jtype::HeapAddr;
use common::utils::cursor::CursorError;
use thiserror::Error;

//TODO: there is a way to not copy-paste macros? refactor later
#[macro_export]
macro_rules! throw_unsupported_operation {
    ($fmt:literal $(, $args:expr)* $(,)?) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::UnsupportedOperationException(format!($fmt $(, $args)*)),
        ))
    };
    ($msg:expr) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::UnsupportedOperationException($msg.into()),
        ))
    };
}

#[macro_export]
macro_rules! throw_arithmetic_exception {
    ($fmt:literal $(, $args:expr)* $(,)?) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::ArithmeticException(format!($fmt $(, $args)*)),
        ))
    };
    ($msg:expr) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::ArithmeticException($msg.into()),
        ))
    };
}

#[macro_export]
macro_rules! throw_negative_array_size_exception {
    ($fmt:literal $(, $args:expr)* $(,)?) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::NegativeArraySizeException(format!($fmt $(, $args)*)),
        ))
    };
    ($msg:expr) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::NegativeArraySizeException($msg.into()),
        ))
    };
}

#[macro_export]
// TODO: can have fixed message template
macro_rules! throw_index_out_of_bounds_exception {
    ($fmt:literal $(, $args:expr)* $(,)?) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::ArrayIndexOutOfBoundsException(format!($fmt $(, $args)*)),
        ))
    };
    ($msg:expr) => {
        Err(JavaExceptionFromJvm::JavaLang(
            JavaLangError::ArrayIndexOutOfBoundsException($msg.into()),
        ))
    };
}

#[derive(Debug, Error)]
pub enum JvmError {
    #[error("LinkageError: {0}")]
    Linkage(#[from] LinkageError),
    #[error(transparent)]
    Cursor(#[from] CursorError),
    #[error("RuntimeConstantPoolError: {0}")]
    RuntimePool(#[from] RuntimePoolError),
    #[error(transparent)]
    ClassLoader(#[from] ClassLoaderErr),
    #[error("MissingAttributeInConstantPoll")]
    MissingAttributeInConstantPoll,
    #[error("ConstantNotFoundInRuntimePool")]
    ConstantNotFoundInRuntimePool,
    #[error("TrailingBytes")]
    TrailingBytes,
    #[error("ClassNotFoundException: {0}")]
    ClassNotFound(String),
    #[error("stack overflow")]
    StackOverflow,
    #[error("Frame stack is empty")]
    FrameStackIsEmpty,
    #[error("Operand stack is empty")]
    OperandStackIsEmpty,
    #[error("OutOfMemory")]
    OutOfMemory,
    #[error("Could not find or load main class {0}")]
    NoMainClassFound(String),
    #[error("NoSuchMethod: {0}")]
    NoSuchMethod(String),
    #[error("NoSuchField: {0}")]
    NoSuchFieldError(String),
    #[error("LocalVariableNotFound: {0}")]
    LocalVariableNotFound(u8),
    #[error("LocalVariableNotInitialized: {0}")]
    LocalVariableNotInitialized(u8),
    #[error("TypeDescriptorErr: {0}")]
    TypeDescriptorErr(#[from] common::TypeDescriptorErr),
    #[error("NullPointerException")]
    NullPointerException,
    #[error("InstructionErr: {0}")]
    InstructionErr(#[from] common::InstructionErr),
    #[error("ClassMirrorIsAlreadyCreated")]
    ClassMirrorIsAlreadyCreated,
    #[error("Method is not expecting to be abstract `{0}`")]
    MethodIsAbstract(String),
    #[error("UnexpectedType: `{0}`")]
    UnexpectedType(String),
    #[error("JavaExceptionThrown: `{0}`")]
    JavaExceptionThrown(HeapAddr),
    #[error("Uninitialized")]
    Uninitialized,
    #[error("WrongHeapAddress: `{0}`")]
    WrongHeapAddress(HeapAddr),
    #[error("TODO map to correct error: `{0}`")]
    Todo(String),
    #[error("JavaLangError: {0}")]
    JavaException(#[from] JavaExceptionFromJvm),
}

// TODO: everything below needs to be refactored
#[derive(Debug, Error)]
pub enum JavaExceptionFromJvm {
    #[error("LinkageError: {0}")]
    JavaLang(JavaLangError),
}

impl JavaExceptionFromJvm {
    pub fn as_reference(&self) -> JavaExceptionReference {
        match self {
            JavaExceptionFromJvm::JavaLang(err) => err.as_reference(),
        }
    }

    pub fn get_message(&self) -> &String {
        match self {
            JavaExceptionFromJvm::JavaLang(err) => err.get_message(),
        }
    }
}

pub struct JavaExceptionReference {
    pub class: &'static str,
    pub name: &'static str,
    pub descriptor: &'static str,
}

#[derive(Debug, Error)]
pub enum JavaLangError {
    #[error("java.lang.ArithmeticException: {0}")]
    ArithmeticException(String),
    #[error("java.lang.UnsupportedOperationException: {0}")]
    UnsupportedOperationException(String),
    #[error("java.lang.ArrayIndexOutOfBoundsException: {0}")]
    ArrayIndexOutOfBoundsException(String),
    #[error("java.lang.NegativeArraySizeException: {0}")]
    NegativeArraySizeException(String),
}

impl JavaLangError {
    pub fn as_reference(&self) -> JavaExceptionReference {
        match self {
            JavaLangError::ArithmeticException(_) => JavaExceptionReference {
                class: "java/lang/ArithmeticException",
                name: "<init>",
                descriptor: "(Ljava/lang/String;)V",
            },
            JavaLangError::UnsupportedOperationException(_) => JavaExceptionReference {
                class: "java/lang/UnsupportedOperationException",
                name: "<init>",
                descriptor: "(Ljava/lang/String;)V",
            },
            JavaLangError::ArrayIndexOutOfBoundsException(_) => JavaExceptionReference {
                class: "java/lang/ArrayIndexOutOfBoundsException",
                name: "<init>",
                descriptor: "(Ljava/lang/String;)V",
            },
            JavaLangError::NegativeArraySizeException(_) => JavaExceptionReference {
                class: "java/lang/NegativeArraySizeException",
                name: "<init>",
                descriptor: "(Ljava/lang/String;)V",
            },
        }
    }

    pub fn get_message(&self) -> &String {
        match self {
            JavaLangError::ArithmeticException(msg) => msg,
            JavaLangError::UnsupportedOperationException(msg) => msg,
            JavaLangError::ArrayIndexOutOfBoundsException(msg) => msg,
            JavaLangError::NegativeArraySizeException(msg) => msg,
        }
    }
}
