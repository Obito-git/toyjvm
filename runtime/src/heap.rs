// TODO: very primitive implementation, ok for right now

use crate::JvmError;
use crate::rt::class::class::Class;
use crate::rt::class::field::Field;
use crate::rt::constant_pool::reference::NameAndTypeReference;
use common::jtype::{HeapAddr, Value};
use std::sync::Arc;
use tracing_log::log::debug;

pub enum HeapObject {
    Instance(ClassInstance),
    String(String),
}

#[derive(Clone)]
pub struct ClassInstance {
    class: Arc<Class>,
    fields: Vec<Value>,
}

impl ClassInstance {
    pub fn get_field(&mut self, index: usize) -> &mut Value {
        self.fields.get_mut(index).expect("invalid field index")
    }

    pub fn class(&self) -> &Arc<Class> {
        &self.class
    }
}

/// https://docs.oracle.com/javase/specs/jvms/se24/html/jvms-2.html#jvms-2.5.3
pub struct Heap {
    objects: Vec<HeapObject>,
}

impl Heap {
    pub fn new() -> Self {
        debug!("Initializing Heap...");
        Self {
            objects: Vec::new(),
        }
    }

    fn push(&mut self, obj: HeapObject) -> HeapAddr {
        let idx = self.objects.len();
        self.objects.push(obj);
        idx
    }

    //TODO: how to handle super classes?
    pub fn alloc_instance(&mut self, class: Arc<Class>) -> HeapAddr {
        let fields = class
            .fields()
            .iter()
            .map(|field| field.descriptor().resolved().get_default_value())
            .collect();
        self.push(HeapObject::Instance(ClassInstance { class, fields }))
    }

    pub fn alloc_string<S: Into<String>>(&mut self, s: S) -> HeapAddr {
        self.push(HeapObject::String(s.into()))
    }

    pub fn get(&self, h: HeapAddr) -> &HeapObject {
        self.objects.get(h).expect("heap: invalid handle (get)")
    }

    pub fn get_instance(&mut self, h: &HeapAddr) -> &ClassInstance {
        let heap_obj = self.get(*h);
        match heap_obj {
            HeapObject::Instance(inst) => inst,
            _ => panic!("get_by_ref called with non-instance HeapObject",),
        }
    }

    pub fn get_instance_field(&mut self, h: &HeapAddr, nat: &NameAndTypeReference) -> &Value {
        let instance = self.get_instance(h);
        let slot = instance.class.get_field_index(nat).unwrap();
        instance.fields.get(slot).unwrap()
    }

    pub fn get_mut(&mut self, h: HeapAddr) -> &mut HeapObject {
        self.objects
            .get_mut(h)
            .expect("heap: invalid handle (get_mut)")
    }

    pub fn write_instance_field(
        &mut self,
        h: HeapAddr,
        field_nat: &NameAndTypeReference,
        val: Value,
    ) -> Result<(), JvmError> {
        match self.get_mut(h) {
            HeapObject::Instance(instance) => {
                let slot = instance.class.get_field_index(field_nat)?;
                instance.fields[slot] = val;
            }
            _ => panic!("heap: write_instance_field on non-instance"),
        }
        Ok(())
    }
}
