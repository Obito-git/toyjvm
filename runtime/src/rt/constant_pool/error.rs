use crate::rt::constant_pool::RuntimeConstantType;
use common::{MethodDescriptorErr, TypeDescriptorErr};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum RuntimePoolError {
    #[error(transparent)]
    MethodDescriptor(#[from] MethodDescriptorErr),
    #[error(transparent)]
    TypeDescriptor(#[from] TypeDescriptorErr),
    #[error("WrongIndex")]
    WrongIndex(u16),
    #[error("TypeError")]
    TypeError(u16, RuntimeConstantType, RuntimeConstantType), //TODO: named args?
    #[error("TryingToAccessUnresolved")]
    TryingToAccessUnresolved(u16, RuntimeConstantType),
}
