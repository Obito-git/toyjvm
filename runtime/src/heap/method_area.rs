use crate::class_loader::ClassLoader;
use crate::rt::class::Class;
use crate::rt::method::Method;
use crate::{ClassId, FieldDescriptorId, MethodId, Symbol, VmConfig};
use common::error::{JvmError, LinkageError};
use common::jtype::Type;
use jclass::ClassFile;
use lasso::{Spur, ThreadedRodeo};
use std::collections::HashMap;
use std::sync::Arc;

pub struct MethodArea {
    bootstrap_class_loader: ClassLoader,
    class_name_to_index: HashMap<Spur, ClassId>,
    classes: Vec<Class>,
    methods: Vec<Method>,
    field_descriptors: Vec<Type>,
    field_descriptors_index: HashMap<Symbol, FieldDescriptorId>,
    pub string_interner: Arc<ThreadedRodeo>,
    constructor_symbols: (Symbol, Symbol),
}

impl MethodArea {
    pub fn new(
        vm_config: &VmConfig,
        string_interner: Arc<ThreadedRodeo>,
    ) -> Result<Self, JvmError> {
        let bootstrap_class_loader = ClassLoader::new(vm_config)?;
        let constructor_symbols = (
            string_interner.get_or_intern("<init>"),
            string_interner.get_or_intern("<clinit>"),
        );

        Ok(Self {
            bootstrap_class_loader,
            class_name_to_index: HashMap::new(),
            classes: Vec::new(),
            methods: Vec::new(),
            field_descriptors: Vec::new(),
            field_descriptors_index: HashMap::new(),
            string_interner,
            constructor_symbols,
        })
    }

    // todo: probably there is better place to put this
    pub fn is_constructor_symbol(&self, name: Symbol) -> bool {
        self.constructor_symbols.0 == name || self.constructor_symbols.1 == name
    }

    fn push_field_descriptor(&mut self, ty: Type) -> FieldDescriptorId {
        self.field_descriptors.push(ty);
        FieldDescriptorId::from_usize(self.field_descriptors.len())
    }

    pub fn get_field_descriptor(&self, id: &FieldDescriptorId) -> &Type {
        &self.field_descriptors[id.to_index()]
    }

    pub fn get_or_new_field_descriptor_id(
        &mut self,
        descriptor: &Symbol,
    ) -> Result<FieldDescriptorId, JvmError> {
        if let Some(field_desc) = self.field_descriptors_index.get(descriptor) {
            return Ok(*field_desc);
        }
        let descriptor_str = self.string_interner.resolve(descriptor);
        let ty = Type::try_from(descriptor_str)?;
        Ok(self.push_field_descriptor(ty))
    }

    pub fn push_method(&mut self, method: Method) -> MethodId {
        self.methods.push(method);
        MethodId::from_usize(self.methods.len())
    }

    pub fn get_method(&self, method_id: &MethodId) -> &Method {
        &self.methods[method_id.to_index()]
    }

    pub fn push_class(&mut self, class: Class) -> ClassId {
        self.classes.push(class);
        ClassId::from_usize(self.classes.len())
    }

    pub fn get_class(&self, class_id: &ClassId) -> &Class {
        &self.classes[class_id.to_index()]
    }

    fn load_class(&mut self, name_sym: Symbol) -> Result<ClassId, JvmError> {
        let data = {
            let name_str = self.string_interner.resolve(&name_sym);
            self.bootstrap_class_loader.load(name_str)?
        };
        let cf = ClassFile::try_from(data).map_err(LinkageError::from)?;
        let super_id = match cf.get_super_class_name() {
            Some(super_name) => {
                let super_name = super_name.unwrap();
                let super_name_sym = self.string_interner.get_or_intern(super_name);
                Some(self.get_class_id_or_load(super_name_sym)?)
            }
            None => None,
        };
        let class_id = Class::load_and_link(cf, self, super_id)?;
        self.class_name_to_index.insert(name_sym, class_id);
        Ok(class_id)
    }

    pub fn get_class_id_or_load(&mut self, name_sym: Symbol) -> Result<ClassId, JvmError> {
        if let Some(class_id) = self.class_name_to_index.get(&name_sym) {
            return Ok(*class_id);
        }
        let class_id = self.load_class(name_sym)?;
        Ok(class_id)
    }
}
