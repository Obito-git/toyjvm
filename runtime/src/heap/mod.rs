// TODO: very primitive implementation, ok for right now

use crate::heap::method_area::MethodArea;
use crate::rt::class::Class;
use crate::{ClassId, throw_index_out_of_bounds_exception};
use common::error::JavaExceptionFromJvm;
use common::error::JavaLangError;
use common::error::JvmError;
use common::jtype::{HeapAddr, Value};
use std::collections::HashMap;
use std::sync::Arc;
use tracing_log::log::debug;

pub mod method_area;

pub enum HeapObject {
    Instance(Instance),
    Array(Instance),
}

#[derive(Clone)]
pub struct Instance {
    class_id: ClassId,
    data: Vec<Value>,
}

impl Instance {
    pub fn new(class_id: ClassId, elements: Vec<Value>) -> Self {
        Self {
            class_id,
            data: elements,
        }
    }

    pub fn class_id(&self) -> &ClassId {
        &self.class_id
    }

    pub fn elements(&self) -> &Vec<Value> {
        &self.data
    }

    pub fn elements_mut(&mut self) -> &mut Vec<Value> {
        &mut self.data
    }

    pub fn length(&self) -> usize {
        self.data.len()
    }

    pub fn get_element(&self, index: i32) -> Result<&Value, JvmError> {
        // TODO: this block is repeated at least 3 times, refactor
        let index = if index >= 0 && (index as usize) < self.data.len() {
            index as usize
        } else {
            throw_index_out_of_bounds_exception!(
                "Index {} out of bounds for length {}",
                index,
                self.data.len()
            )?
        };
        Ok(&self.data[index])
    }

    pub fn get_element_mut(&mut self, index: i32) -> Result<&mut Value, JvmError> {
        // TODO: this block is repeated at least 3 times, refactor
        let index = if index >= 0 && (index as usize) < self.data.len() {
            index as usize
        } else {
            throw_index_out_of_bounds_exception!(
                "Index {} out of bounds for length {}",
                index,
                self.data.len()
            )?
        };
        Ok(&mut self.data[index])
    }
}

// TODO: probably should be different
// right now to avoid self.alloc_instance(&self.class_class)?; don't want to clone to avoid using self self
enum AllocClass<'a> {
    String,
    Char,
    Class,
    Other(&'a Arc<Class>),
}

/// https://docs.oracle.com/javase/specs/jvms/se24/html/jvms-2.html#jvms-2.5.3
pub struct Heap {
    objects: Vec<HeapObject>,
    string_pool: HashMap<String, HeapAddr>,
    // TODO: find a better way to expose MethodArea functionality
    string_class: Arc<Class>,
    char_class: Arc<Class>,
    class_class: Arc<Class>,
    mirrors: HashMap<HeapAddr, Arc<Class>>,
    primitives: HashMap<HeapAddr, HeapAddr>,
}

impl Heap {
    pub fn new(method_area: &mut MethodArea) -> Result<Self, JvmError> {
        debug!("Creating Heap...");
        let char_class = method_area.get_class_or_load_by_name("[C")?.clone();
        let string_class = method_area
            .get_class_or_load_by_name("java/lang/String")?
            .clone();
        let class_class = method_area
            .get_class_or_load_by_name("java/lang/Class")?
            .clone();
        Ok(Self {
            string_pool: HashMap::new(),
            objects: Vec::new(),
            mirrors: HashMap::new(),
            primitives: HashMap::new(),
            char_class,
            string_class,
            class_class,
        })
    }

    pub fn get_class_id(&self, h: &HeapAddr) -> ClassId {
        match self.get(*h).unwrap() {
            HeapObject::Instance(inst) => inst.class_id,
            HeapObject::Array(arr) => arr.class_id,
        }
    }

    fn push(&mut self, obj: HeapObject) -> HeapAddr {
        let idx = self.objects.len();
        self.objects.push(obj);
        idx
    }

    pub fn alloc_array(&mut self, class: &Arc<Class>, length: usize) -> Result<HeapAddr, JvmError> {
        let default_value = if let Some(primitive_type) = class.primitive() {
            Value::from(&primitive_type)
        } else {
            Value::Null
        };
        let elements = vec![default_value; length];
        Ok(self.push(HeapObject::Array(Instance::new(*class.id(), elements))))
    }

    pub fn alloc_array_with_value(
        &mut self,
        class: &Arc<Class>,
        length: usize,
        value: Value,
    ) -> Result<HeapAddr, JvmError> {
        let elements = vec![value; length];
        Ok(self.push(HeapObject::Array(Instance::new(*class.id(), elements))))
    }

    fn create_default_fields(class: &Arc<Class>) -> Vec<Value> {
        let mut fields = Vec::with_capacity(class.fields().len());

        let mut cur_class = Some(class);

        while let Some(c) = cur_class {
            for field in c.fields() {
                fields.push(field.descriptor().resolved().get_default_value());
            }
            cur_class = c.super_class();
        }

        fields
    }

    fn alloc_instance_internal(&mut self, class: AllocClass) -> Result<HeapAddr, JvmError> {
        let class = match class {
            AllocClass::String => &self.string_class,
            AllocClass::Char => &self.char_class,
            AllocClass::Class => &self.class_class,
            AllocClass::Other(c) => c,
        };
        let fields = Self::create_default_fields(class);
        Ok(self.push(HeapObject::Instance(Instance {
            class_id: *class.id(),
            data: fields,
        })))
    }

    pub fn alloc_instance(&mut self, class: &Arc<Class>) -> Result<HeapAddr, JvmError> {
        self.alloc_instance_internal(AllocClass::Other(class))
    }

    pub fn get_or_new_string(&mut self, value: &str) -> HeapAddr {
        debug!("Getting or creating string in pool: {}", value);
        if let Some(&h) = self.string_pool.get(value) {
            debug!("String found in pool: {}", value);
            return h;
        }
        debug!("String not found in pool. Creating new one: {}", value);
        let h = self.alloc_string(value);
        self.string_pool.insert(value.to_string(), h);
        h
    }

    fn alloc_string(&mut self, s: &str) -> HeapAddr {
        let mut fields = Self::create_default_fields(&self.string_class);

        let chars = s.chars().map(|c| Value::Integer(c as i32)).collect();
        let value = self.push(HeapObject::Array(Instance::new(
            *self.char_class.id(),
            chars,
        )));

        // The "value" field is always the first field in java.lang.String
        fields[0] = Value::Ref(value);

        self.push(HeapObject::Instance(Instance {
            class_id: *self.string_class.id(),
            data: fields,
        }))
    }

    // TODO: return Result and handle errors
    pub fn get(&self, h: HeapAddr) -> Option<&HeapObject> {
        self.objects.get(h)
    }

    pub fn get_instance(&self, h: &HeapAddr) -> Result<&Instance, JvmError> {
        let heap_obj = self.get(*h).unwrap();
        match heap_obj {
            HeapObject::Instance(inst) => Ok(inst),
            _ => Err(JvmError::Todo(
                "get_by_ref called with non-instance HeapObject".to_string(),
            )),
        }
    }

    pub fn get_instance_mut(&mut self, h: &HeapAddr) -> &mut Instance {
        let heap_obj = self.get_mut(*h);
        match heap_obj {
            HeapObject::Instance(inst) => inst,
            _ => panic!("get_by_ref called with non-instance HeapObject",),
        }
    }

    pub fn get_array(&self, h: &HeapAddr) -> &Instance {
        let heap_obj = self.get(*h).unwrap();
        match heap_obj {
            HeapObject::Array(arr) => arr,
            _ => panic!("get_array called with non-array HeapObject",),
        }
    }

    pub fn get_instance_field(
        &mut self,
        addr: &HeapAddr,
        offset: usize,
    ) -> Result<&Value, JvmError> {
        let instance = self.get_instance(addr)?;
        instance
            .data
            .get(offset)
            .ok_or(JvmError::Todo("invalid field index".to_string()))
    }

    pub fn get_mut(&mut self, h: HeapAddr) -> &mut HeapObject {
        self.objects
            .get_mut(h)
            .expect("heap: invalid handle (get_mut)")
    }

    pub fn addr_is_instance(&self, h: &HeapAddr) -> bool {
        matches!(self.get(*h), Some(HeapObject::Instance(_)))
    }

    //TODO: design it lightweight
    pub fn get_string(&self, h: HeapAddr) -> Result<String, JvmError> {
        let instance = self.get_instance(&h)?;
        let value_field = instance.get_element(0)?; // "value" field is always the first field in java.lang.String
        let array_addr = match value_field {
            Value::Ref(addr) => *addr,
            _ => {
                return Err(JvmError::JavaException(JavaExceptionFromJvm::JavaLang(
                    JavaLangError::NullPointerException,
                )));
            }
        };
        let char_array = self.get_array(&array_addr);
        let chars: String = char_array
            .elements()
            .iter()
            .map(|v| match v {
                Value::Integer(i) => std::char::from_u32(*i as u32).unwrap_or('*'),
                _ => '*',
            })
            .collect();
        Ok(chars)
    }

    pub fn write_array_element(
        &mut self,
        h: HeapAddr,
        index: i32,
        val: Value,
    ) -> Result<(), JvmError> {
        match self.get_mut(h) {
            HeapObject::Array(array) => {
                let index = if index >= 0 && (index as usize) < array.length() {
                    index as usize
                } else {
                    throw_index_out_of_bounds_exception!(
                        "Index {} out of bounds for length {}",
                        index,
                        array.length()
                    )?
                };
                array.data[index] = val;
            }
            _ => panic!("heap: write_array_element on non-array"),
        }
        Ok(())
    }

    pub fn write_instance_field(
        &mut self,
        h: HeapAddr,
        offset: usize,
        val: Value,
    ) -> Result<(), JvmError> {
        match self.get_mut(h) {
            HeapObject::Instance(instance) => {
                if offset >= instance.data.len() {
                    return Err(JvmError::Todo("invalid field index".to_string()));
                }
                instance.data[offset] = val;
            }
            _ => Err(JvmError::Todo(
                "heap: write_instance_field on non-instance".to_string(),
            ))?,
        }
        Ok(())
    }

    pub fn clone_object(&mut self, h: HeapAddr) -> HeapAddr {
        let obj = self.get(h).expect("heap: invalid handle (clone_object)");
        match obj {
            HeapObject::Instance(inst) => {
                let new_fields = inst.data.clone();
                let new_instance = Instance {
                    class_id: inst.class_id,
                    data: new_fields,
                };
                self.push(HeapObject::Instance(new_instance))
            }
            HeapObject::Array(arr) => {
                let new_elements = arr.data.clone();
                let new_array = Instance {
                    class_id: arr.class_id,
                    data: new_elements,
                };
                self.push(HeapObject::Array(new_array))
            }
        }
    }

    pub fn get_class_by_mirror(&self, mirror: &HeapAddr) -> Option<&Arc<Class>> {
        self.mirrors.get(mirror)
    }

    pub(crate) fn get_mirror_addr(
        &mut self,
        target_class: &Arc<Class>,
    ) -> Result<HeapAddr, JvmError> {
        if let Some(mirror) = target_class.mirror() {
            return Ok(mirror);
        }
        let mirror = self.alloc_instance_internal(AllocClass::Class)?;
        target_class.set_mirror(mirror)?;
        self.mirrors.insert(mirror, target_class.clone());
        Ok(mirror)
    }

    pub(crate) fn get_primitive_mirror_addr(
        &mut self,
        name: &HeapAddr,
    ) -> Result<HeapAddr, JvmError> {
        if let Some(addr) = self.primitives.get(name) {
            Ok(*addr)
        } else {
            let mirror_addr = self.alloc_instance_internal(AllocClass::Class)?;
            self.primitives.insert(*name, mirror_addr);
            Ok(mirror_addr)
        }
    }

    pub fn addr_is_primitive(&self, addr: &HeapAddr) -> bool {
        self.primitives.values().any(|v| v == addr)
    }
}
