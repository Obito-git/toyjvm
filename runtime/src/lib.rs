use crate::error::JvmError;
use crate::heap::Heap;
use crate::interpreter::Interpreter;
use crate::method_area::MethodArea;
use crate::native::NativeRegistry;
use crate::stack::FrameStack;
use std::cell::RefCell;
use std::rc::Rc;
use tracing_log::log::debug;

mod class_loader;
pub mod error;
mod heap;
mod interpreter;
mod method_area;
mod native;
pub mod rt;
pub mod stack;

pub type ClassId = usize;

#[derive(Debug)]
pub struct VmConfig {
    pub home: String,
    pub version: String,
    pub class_path: Vec<String>,
    pub initial_heap_size: usize,
    pub max_heap_size: usize,
    pub frame_stack_size: usize,
    pub operand_stack_size: usize,
}

//TODO: make it better
impl VmConfig {
    pub fn validate(&self) {
        if self.version != "24.0.2" {
            panic!(
                "Unsupported Java version: {}. Only 24.0.2 is supported.",
                self.version
            );
        }
    }
}

pub struct VirtualMachine {
    config: VmConfig,
    method_area: MethodArea,
    heap: Rc<RefCell<Heap>>,
    native_registry: NativeRegistry,
    frame_stack: FrameStack,
}

impl VirtualMachine {
    pub fn new(config: VmConfig) -> Result<Self, JvmError> {
        config.validate();
        let heap = Rc::new(RefCell::new(Heap::new()));
        let mut method_area = MethodArea::new(&config, heap.clone())?;
        debug!("Preloading java/lang/String...");
        let string_class = method_area.get_class("java/lang/String")?;
        heap.borrow().initialize(string_class);

        let native_registry = NativeRegistry::new();
        Ok(Self {
            frame_stack: FrameStack::new(&config),
            config,
            method_area,
            heap,
            native_registry,
        })
    }

    pub fn method_area(&mut self) -> &mut MethodArea {
        &mut self.method_area
    }

    pub fn heap(&self) -> Rc<RefCell<Heap>> {
        self.heap.clone()
    }
}

pub fn start(main_class: Vec<u8>, config: VmConfig) -> Result<(), JvmError> {
    debug!("Starting VM with config: {:?}", config);

    let vm = VirtualMachine::new(config)?;

    let mut interpreter = Interpreter::new(vm);
    interpreter.start(main_class)
}

#[cfg(test)]
impl serde::Serialize for VirtualMachine {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        use serde::ser::SerializeStruct;
        use std::ops::Deref;

        let mut state = serializer.serialize_struct("VirtualMachine", 3)?;
        state.serialize_field("heap", self.heap.borrow().deref())?;
        state.end()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::with_settings;
    use rstest::rstest;
    use std::fs;
    use std::path::{Path, PathBuf};

    const DISPLAY_SNAPSHOT_PATH: &str = "../snapshots";

    fn to_snapshot_name(path: &Path) -> String {
        let mut iter = path.iter().map(|s| s.to_string_lossy().to_string());
        for seg in iter.by_ref() {
            if seg == "test-classes" {
                break;
            }
        }
        let tail: Vec<String> = iter.collect();
        tail.join("-")
    }

    fn vm_config() -> VmConfig {
        let java_home = std::env::var("JAVA_HOME").expect("JAVA_HOME must be set for tests");

        VmConfig {
            home: java_home,
            version: "24.0.2".to_string(),
            class_path: vec![],
            initial_heap_size: 16 * 1024 * 1024,
            max_heap_size: 64 * 1024 * 1024,
            frame_stack_size: 256,
            operand_stack_size: 256,
        }
    }

    fn main_base_package_dir(path: &PathBuf) -> String {
        let mut dir = path.parent().expect("file should have a parent dir");

        dir = dir.parent().expect("should have parent dir");
        dir = dir.parent().expect("should have parent dir");

        dir.to_string_lossy().to_string()
    }

    #[rstest]
    #[trace]
    // Requires `testdata/compile-fixtures.sh` to be run to generate the .class files
    fn test_display(
        #[base_dir = "../target/test-classes/custom"]
        #[files("**/*Main.class")]
        path: PathBuf,
    ) {
        // Given
        let bytes = fs::read(&path).unwrap_or_else(|_| panic!("Can't read file {:?}", path));
        let mut base_config = vm_config();
        base_config.class_path.push(main_base_package_dir(&path));
        let vm = VirtualMachine::new(base_config).unwrap();
        let mut interpreter = Interpreter::new(vm);

        // When
        interpreter
            .start(bytes)
            .unwrap_or_else(|e| panic!("Interpreter failed for {:?} with error: {:?}", path, e));

        // Then
        with_settings!(
            {
                snapshot_path => DISPLAY_SNAPSHOT_PATH,
                prepend_module_to_snapshot => false,
            },
            {
                insta::assert_yaml_snapshot!(to_snapshot_name(&path), &interpreter);
            }
        );
    }
}
