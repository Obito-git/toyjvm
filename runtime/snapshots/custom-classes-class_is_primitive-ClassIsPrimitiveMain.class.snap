---
source: runtime/src/lib.rs
expression: "&interpreter"
---
vm:
  heap:
    objects:
      - address: 0
        type: Array
        class: "[C"
        elements:
          - Integer: 105
          - Integer: 110
          - Integer: 116
      - address: 1
        type: Instance
        value:
          class: java/lang/String
          fields:
            - Ref: 0
            - Integer: 0
            - Integer: 0
            - Integer: 0
      - address: 2
        type: Array
        class: java/lang/Class
        elements: []
      - address: 3
        type: Array
        class: java/io/ObjectStreamField
        elements: []
      - address: 4
        type: Instance
        value:
          class: java/lang/Class
          fields:
            - "Null"
            - "Null"
            - "Null"
            - "Null"
            - "Null"
            - "Null"
            - "Null"
            - "Null"
            - "Null"
            - Integer: 0
            - "Null"
            - "Null"
            - "Null"
            - "Null"
            - "Null"
            - "Null"
      - address: 5
        type: Array
        class: "[C"
        elements:
          - Integer: 48
          - Integer: 49
          - Integer: 50
          - Integer: 51
          - Integer: 52
          - Integer: 53
          - Integer: 54
          - Integer: 55
          - Integer: 56
          - Integer: 57
          - Integer: 97
          - Integer: 98
          - Integer: 99
          - Integer: 100
          - Integer: 101
          - Integer: 102
          - Integer: 103
          - Integer: 104
          - Integer: 105
          - Integer: 106
          - Integer: 107
          - Integer: 108
          - Integer: 109
          - Integer: 110
          - Integer: 111
          - Integer: 112
          - Integer: 113
          - Integer: 114
          - Integer: 115
          - Integer: 116
          - Integer: 117
          - Integer: 118
          - Integer: 119
          - Integer: 120
          - Integer: 121
          - Integer: 122
    string_pool:
      - string: int
        address: 1
last_pop_frame:
  locals:
    - ~
    - Ref: 4
    - Integer: 1
  operands: []
  pc: 10
